sdlib/d/gc/util.d

```diff
@@ -288,3 +292,46 @@ unittest alignUp {
 		}
 	}
 }
+
+/**
+ * A 'packedU15' represents a 15-bit unsigned integer residing in one or two bytes.
+ * It is always written in 'big-endian' orientation, with the least-significant
+ * byte (always present) occupying position 'loc + 1', and the most-significant
+ * byte (may or may not be present) occupying position 'loc'.
+ * The least-significant bit of the upper byte indicates the presence of the
+ * lower byte; its remaining 7 bits store the least-significant 7 bits of the
+ * value. The lower byte, if present, stores the most significant 8 bits.
+ */
+
+ushort readPackedU15(void* loc) {
+	// FIXME: detect that we're actually on a little-endian machine:
+	auto data = swapEndian(*(cast(ushort*) loc));
+	auto mask = 0x7f | -(data & 1);
+	return (data >>> 1) & mask;
+}
+
+void writePackedU15(void* loc, ushort x) {
+	auto locptr = (cast(ushort*) loc);
+	// If writing a one-byte packedU15, don't disturb the unused byte:
+	ushort mask = 0xff00 | (0xff & ((0x7f - x) >> 16));
```

Comment:
I believe this is overly complex. The 0xff mask on the rhs for instance is completely superfluous.

`ushort mask = 0xff00 | -(x > 0x7f)` seems equivalent (you might need a cast). You might benefit from storing the comparison down the road considering it reused in the form of `mask & 1`, which is completely opaque to the reader and more work for the computer to do.

But ultimately, and I might be wrong there, but it seems like you'd be better off computing the current value, then the delta, and the delta and set the last bit. Something like this (untested).

```d
auto data = loadBigEndian(cast(ushort*) loc);
auto mask = 0xff | -(data & 1);
auto current = data & mask;
auto delta = (x << 1) - current;
auto new = (data + delta) | (x > 0x7f);
storeBigEndian(loc, new);
```

Reply:
Maybe even the compared can be lifted into the delta computation, as in:

```d
auto data = loadBigEndian(loc);
auto mask = 0xff | -(data & 1);
auto current = data & mask;
auto base = (x << 1) | (x > 0x7f);
auto delta = base - current;
auto value = data + delta;
storeBigEndian(loc, value);
```

Still untested, but it seems like a variation of this should work.

Reply:
I played around a bit, the best I can find are:
```d
void v1(ushort* ptr, ushort x) {
    auto base = (x << 1) | (x > 0x7f);
    auto mask = -(x > 0x7f) | 0xff00;

    import core.bitop;
    auto value = byteswap(base & ushort.max);

    auto current = *ptr;
    ushort delta = (current ^ value) & mask;
    *ptr = current ^ delta;
}

void v2(ushort* ptr, ushort x) {
    auto base = (x << 1) | (x > 0x7f);
    auto mask = -(x > 0x7f) | 0xff;

    auto current = loadBigEndian(ptr);
    auto delta = (current ^ base) & mask;
    auto value = current ^ delta;

    storeBigEndian(ptr, value & ushort.max);
}
```

v2 is platform agnostic, v1 isn't. Both are better than both of our solutions.